### PR TITLE
[#42] [#43] [Backend] [Integrate] As a user, I can use the app without issues even if my token is expired

### LINKED ISSUE #42: [Integrate] As a user, I can use the app without issues even if my token is expired
## Why

As the access token has an expiry date, so we need to refresh it to keep the user session

## Acceptance Criteria

- Refresh the access token when we have `401` status from any API response

### LINKED ISSUE #43: [Backend] As a user, I can use the app without issues even if my token is expired
## Why

When a user finishes submitting their survey, we need to show a screen to thank them for spending their time doing the survey

## Acceptance Criteria

Call `POST /api/v1/oauth/token` endpoint with the following body:

```
{
    "grant_type": "refresh_token",
    "refresh_token": "{{refresh_token}}"",
    "client_id": "{{client_id}}",
    "client_secret": "{{client_secret}}"
}
```

## Resources

Refresh Token: https://nimblehq.postman.co/workspace/Nimble~9daf6b25-882e-4ad0-8299-95529e2883ff/request/11835486-5f7ef181-ab43-43c2-8683-8b487e07cc56

### PR DESCRIPTION
**Pull request by markgravity**
https://github.com/markgravity/kmm-ic/issues/42
https://github.com/markgravity/kmm-ic/issues/43

## What happened 👀

- Add `AuthService.refreshToken` function
- Integrate to refresh token in `ApiService`

## Insight 📝

n/a

## Proof Of Work 📹
CleanShot 2023-05-16 at 10 24 09

### CODE REVIEW

File: shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/data/service/api/ApiService.kt [hunk 1]
```diff
@@ -60,6 +65,25 @@ internal class ApiService(requiresAuthentication: Boolean = false): KoinComponen
                                 BearerTokens(accessToken, refreshToken)
                             }
                         }
+
+                        sendWithoutRequest { request ->
+                            request.url.host == Url(BuildKonfig.BASE_URL).host
+                        }
+
+                        refreshTokens {
+                            print(this)
```
Comment (AVI5HEK) [inline shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/data/service/api/ApiService.kt]: We should remove the print call.

**Review by AVI5HEK — COMMENTED** [review]
The rest LGTM.

File: shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/data/service/api/ApiService.kt [hunk 2]
```diff
@@ -60,6 +65,25 @@ internal class ApiService(requiresAuthentication: Boolean = false): KoinComponen
...
+                            authService
+                                .refreshToken(
+                                    AuthRefreshTokenParams(
+                                        refreshToken = oldTokens?.refreshToken ?: ""
```
Comment (phongvhd93) [inline shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/data/service/api/ApiService.kt]: refreshToken = oldTokens?.refreshToken.orEmpty()

Comment (markgravity) [reply shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/data/service/api/ApiService.kt]: @phongvhd93 updated at [b62ae69](https://github.com/markgravity/kmm-ic/pull/105/commits/b62ae69001d7118d72ac27657189d833b5ef7ffe)

File: shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/data/service/api/ApiService.kt (hunk 1 above)
Comment (markgravity) [reply shared/src/commonMain/kotlin/co/nimblehq/mark/kmmic/data/service/api/ApiService.kt]: @phongvhd93 updated at [b62ae69](https://github.com/markgravity/kmm-ic/pull/105/commits/b62ae69001d7118d72ac27657189d833b5ef7ffe)

**Review by phongvhd93 — APPROVED** [review]

**Review by AVI5HEK — APPROVED** [review]

**Review by suho — APPROVED** [review]

**Pull request merged**